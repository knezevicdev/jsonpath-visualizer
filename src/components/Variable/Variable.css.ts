import styled from 'styled-components';

export const Wrapper = styled.div`
  margin-left: 20px;
`;

export const Name = styled.span`
  display: inline-block;
  color: green;
  letter-spacing: 0.5px;
  font-style: none;
  vertical-align: top;
  opacity: 0.8;

  &:hover {
    opacity: 1;
  }
`;

export const ValueWrapper = styled.span`
  display: inline-block;
  padding-right: 6px;
  position: relative;
`;
