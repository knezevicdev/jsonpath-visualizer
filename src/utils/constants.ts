export const ARRAY_GROUP_LIMIT = 100;
