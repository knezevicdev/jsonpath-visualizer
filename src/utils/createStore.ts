export type TJson = null | Record<string, unknown> | any[];

export const createStore = () => {
  return {
    json: null as TJson,
    jsonPath: '' as string,
  };
};

export type TStore = ReturnType<typeof createStore>;
