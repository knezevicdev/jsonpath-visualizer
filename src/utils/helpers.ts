import { transform as _transform, isObject as _isObject, mapKeys as _mapKeys, assign as _assign } from 'lodash';
import { JSONPath } from 'jsonpath-plus';

import { TJson } from './store';

// eslint-disable-next-line
export const toType = (obj: any): string => {
  let type = getType(obj);
  if (type === 'number') {
    if (isNaN(obj)) {
      type = 'nan';
    } else if ((obj | 0) != obj) {
      //bitwise OR produces integers
      type = 'float';
    } else {
      type = 'integer';
    }
  }
  return type;
};

//source: http://stackoverflow.com/questions/7390426/better-way-to-get-type-of-a-javascript-variable/7390612#7390612
const getType = (obj: any): string => {
  return {}.toString
    .call(obj)
    .match(/\s([a-zA-Z]+)/)[1]
    .toLowerCase();
};

// eslint-disable-next-line
export const objectPaths = (obj: any): Record<string, unknown> => {
  return _transform(
    obj,
    (result: Record<string, unknown>, value: unknown, key: string) => {
      if (_isObject(value)) {
        const flatMap = _mapKeys(objectPaths(value), function (_mvalue, mkey) {
          return `${key}.${mkey}`;
        });

        _assign(result, flatMap);
      } else {
        result[String(key)] = false;
      }

      return result;
    },
    {},
  );
};

export const matchPaths = (json: TJson, path: string): string[] => {
  const paths = JSONPath({
    path,
    json,
    resultType: 'pointer',
  }).map((pointer: string) => pointer.replace(/\//g, '.').substring(1));

  if (paths.includes('')) {
    return [''];
  }

  return paths.reduce((matchedPaths: string[], path: string) => {
    for (const matchedPath in matchedPaths) {
      if (path.includes(matchedPath)) {
        return matchedPaths;
      }
    }

    return [...matchedPaths, path];
  }, []);
};
