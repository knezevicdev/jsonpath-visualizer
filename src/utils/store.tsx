import React, { ComponentType } from 'react';
import { useLocalStore } from 'mobx-react';
import { reaction, comparer, keys, remove } from 'mobx';

import { matchPaths } from 'utils/helpers';

export type TJson = null | Record<string, unknown> | any[];

const storeContext = React.createContext<Record<string, any> | null>(null);

export const StoreProvider: React.FC = ({ children }) => {
  const store = useLocalStore<Record<string, any>>(() => ({
    json: null as TJson,
    jsonPath: '' as string,
    jsonUploadTime: 0 as number,
    matchedPaths: [] as string[],
  }));

  reaction(
    () => store.jsonPath,
    (jsonPath) => {
      const matchedPaths = matchPaths(store.json, jsonPath);

      store.matchedPaths.forEach((oldMatchedPath: string) => {
        if (!matchedPaths.includes(oldMatchedPath)) {
          store[oldMatchedPath] = false;
        }
      });

      matchedPaths.forEach((matchedPath: string) => {
        if (!store[matchedPath]) {
          store[matchedPath] = true;
        }
      });

      store.matchedPaths = matchedPaths;
    },
  );

  reaction(
    () => store.json,
    () => {
      keys(store).forEach((key: string) => {
        if (!['json', 'jsonPath', 'matchedPaths', 'jsonUploadTime'].includes(key)) {
          remove(store, key);
        }
      });

      store.matchedPaths = [];
      store.jsonPath = '';
    },
    { equals: comparer.shallow },
  );

  return <storeContext.Provider value={store}>{children}</storeContext.Provider>;
};

export const withStore = (WrappedComponent: ComponentType): React.FC => {
  const StoreHOC: React.FC = (props) => {
    return (
      <StoreProvider>
        <WrappedComponent {...props} />
      </StoreProvider>
    );
  };

  return StoreHOC;
};

export const useStore = (): Record<string, any> => {
  const store = React.useContext(storeContext);
  if (!store) {
    throw new Error('useStore must be used within a StoreProvider.');
  }
  return store;
};
