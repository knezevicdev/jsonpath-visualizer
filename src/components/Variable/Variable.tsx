import React, { FunctionComponent, useState, useEffect } from 'react';
import PropTypes from 'prop-types';

import { useStore } from 'utils/store';
import { fieldValueProps } from 'styled/helpers';
import { Wrapper, FieldName, FieldValue, FieldValueProps } from 'styled/components';

import { observer } from 'mobx-react';

type VariableProps = {
  name: string;
  value: unknown | null | undefined;
  type: string;
  path: string;
};

const Variable: FunctionComponent<VariableProps> = ({ type, value, name, path }) => {
  const { [path]: matched } = useStore();
  const [props, setProps] = useState<FieldValueProps>(fieldValueProps('orange'));

  useEffect(() => {
    switch (type) {
      case 'string':
        setProps(fieldValueProps('orange'));
        break;
      case 'integer':
        setProps(fieldValueProps('brown'));
        break;
      case 'float':
        setProps(fieldValueProps('green'));
        break;
      case 'boolean':
        setProps(fieldValueProps('purple'));
        break;
      case 'function':
        setProps(fieldValueProps('orange'));
        break;
      case 'null':
        setProps(fieldValueProps('rajah', true));
        break;
      case 'nan':
        setProps(fieldValueProps('pink', true));
        break;
      case 'undefined':
        setProps(fieldValueProps('white', true));
        break;
      case 'date':
        setProps(fieldValueProps('aqua'));
        break;
      case 'regexp':
        setProps(fieldValueProps('aqua'));
        break;
      default:
        // catch-all for types that weren't anticipated
        setProps(fieldValueProps('orange'));
    }
  }, [type]);

  return (
    <Wrapper>
      <FieldName>{name}:&nbsp;</FieldName>
      <FieldValue data-matched={matched} {...props}>
        {type === 'string' && '"'}
        {String(value)}
        {type === 'string' && '"'}
      </FieldValue>
    </Wrapper>
  );
};

Variable.propTypes = {
  name: PropTypes.string.isRequired,
  value: PropTypes.any,
  type: PropTypes.string.isRequired,
  path: PropTypes.string.isRequired,
};

export default observer(Variable);
