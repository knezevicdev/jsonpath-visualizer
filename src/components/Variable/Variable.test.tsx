import React from 'react';
import { shallow } from 'enzyme';

import { StoreProvider } from 'utils/store';

import Variable from './Variable';

describe('Variable', () => {
  it('renders snapshots', () => {
    const wrapper = shallow(
      <StoreProvider>
        <Variable name="" path="" value="" type="string" />
      </StoreProvider>,
    );
    expect(wrapper).toMatchSnapshot();
  });
});
