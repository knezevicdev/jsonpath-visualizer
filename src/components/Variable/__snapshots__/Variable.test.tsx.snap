// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Variable renders snapshots 1`] = `
<ContextProvider
  value={
    Object {
      "json": null,
      "jsonPath": "",
      "jsonUploadTime": 0,
      "matchedPaths": Array [],
    }
  }
>
  <Variable
    name=""
    path=""
    type="string"
    value=""
  />
</ContextProvider>
`;
